sklearn/utils/_array_api.py

```diff
@@ -463,14 +468,25 @@ def _weighted_sum(sample_score, sample_weight, normalize=False, xp=None):
         return float(numpy.average(sample_score_np, weights=sample_weight_np))
 
     if not xp.isdtype(sample_score.dtype, "real floating"):
-        # We move to cpu device ahead of time since certain devices may not support
-        # float64, but we want the same precision for all devices and namespaces.
-        sample_score = xp.astype(xp.asarray(sample_score, device="cpu"), xp.float64)
+        # The MPS device does not support float64
+        if (
+            xp.__name__ in {"array_api_compat.torch", "torch"}
+            and device(sample_score).type == "mps"
```

Comment:
I would rather avoid introducing library specific code path to code related to our use of the Array API in scikit-learn code if we could.

I think we should factor this logic in a dedicated helper such as:

```python
def _max_precision_float_dtype(xp, device):
    # temporary hack while waiting for a proper inspection API, see:
    # https://github.com/data-apis/array-api/issues/640
    if xp.__name__ in {"array_api_compat.torch", "torch"} and device.type == "mps":
        return xp.float32
    else:
        return xp.float64
```


If we really need this, I think we should

Reply:
Done